MassSpectrometry/MsDataFile.cs

```diff
@@ -83,63 +83,175 @@ public static int TopNpeakHelper(ref double[] intensities, ref double[] mArray,
+                    for (int i = 0; i < Math.Min(Temp.Count, filteringParams.NumberOfPeaksToKeepPerWindow.Value); i++)
+                    {
+                        if (WindowMaxNormalizationToValue.HasValue)
+                        {
+                            double normalizedIntensity = Temp[i].Item2 / maxIntensity * (double)WindowMaxNormalizationToValue.Value;
+                            mzIntensites_reduced.Add(new Tuple<double, double>(Temp[i].Item1, normalizedIntensity));
+                        }
+                        else
+                        {
+                            mzIntensites_reduced.Add(Temp[i]);
+                        }
+                    }
+                }
+            }
+
+            // convert merged results to array and sort by m/z
+            mArray = mzIntensites_reduced.Select(i => i.Item1).ToArray();
+            intensities = mzIntensites_reduced.Select(i => i.Item2).ToArray();
+
+            Array.Sort(mArray, intensities);
+        }
+
+        public static void XCorrPrePreprocessing(ref double[] intensities, ref double[] mArray, double scanRangeMinMz, double scanRangeMaxMz, double precursorMz, double precursorDiscardRange = 1.5, double discreteMassBin = 1.0005079, double percentMaxThreshold = 5)
+        {
+            Array.Sort(mArray, intensities);
+            int numberOfWindows = (int)Math.Round((scanRangeMaxMz - scanRangeMinMz + discreteMassBin) / discreteMassBin, 0);
+
+            FilteringParams firstFilter = new FilteringParams(1, percentMaxThreshold / 100, numberOfWindows, false, false);
 
-            // this loop breaks a scan up into "windows" (e.g., a scan with 100 peaks
-            // divided into 10 windows would have 10 peaks per window) and takes the top N peaks per window.
-            // the results of each trimmed window are concatenated into mzResults and intensityResults
-            for (int i = 0; i < filteringParams.NumberOfWindows; i++)
+            for (int i = 0; i < mArray.Length; i++)
             {
-                // make the last window end at the end of the spectrum
-                // this is to prevent rounding errors in windowSize from cutting off the end of the spectrum
-                if (i == filteringParams.NumberOfWindows - 1)
+                if (intensities[i] != 0)
                 {
-                    windowPeakIndexMaximum = intensities.Length - 1;
+                    intensities[i] = Math.Sqrt(intensities[i]);
                 }
+            }
 
-                // avoid index out of range problems
-                if (windowPeakIndexMinimum >= intensities.Length)
+            //set intensity of precursor ions to zero
+            for (int i = 0; i < mArray.Length; i++)
+            {
+                if (mArray[i] > (precursorMz - precursorDiscardRange) && mArray[i] < (precursorMz + precursorDiscardRange))
                 {
-                    break;
+                    intensities[i] = 0;
```

Comment:
So the peak stays, but the intensity is now zero? This will still be counted as a peak match in searches. We should remove the peak from the arrays.

Reply:
for xcorr, there is no peak counting, just dot product. multiplying number times something exponentially close to zero wont budge things much.

Reply:
The method is designed to broadly filter peaks in spectra, not just for xcorr. If we're removing a peak, then we should really remove it.

Reply:
XCorrPrePreprocessing is really only for Xcorr. I don't think there are any zero peaks added in the standard WindowModeHelper, which is the primary filter.